extensions/oplogPopulator/modules/Allocator.js

```diff
@@ -60,7 +55,16 @@ class Allocator {
      * listening to the bucket
      */
     has(bucket) {
-        return this._bucketsToConnectors.has(bucket);
+        return this._connectorsManager.connectors.some(conn => conn.bucketsInUse.has(bucket));
```

Comment:
* this is inefficient, should keep the map
* it is the allocator's job to allocate, it should own it's data (respect levels of abstraction)

Reply:
I did not see a simple solution for this. To effectively detect if a connector has a bucket, we must use the "inUse" Set, **not** the in-memory Set. But, due to the current architecture, we have:
- The Allocator that uses private variables from the ConnectorsManager to access the list of connector, and to create them
- The ConnectorsManager that creates the Connectors and periodically update the pipeline
- The Connector that holds the information about whether the bucket is in use or not

As a result, the Allocator has no way to know, using simple maps, if a bucket is currently in use or not: we consider it in use as soon as it is added, but it stops being in use after 0-5s once we remove it from the in-memory set. This operation being done outside of the Allocator, having a Map would require breaking the abstraction levels, or keeping all the buckets in memory forever...

---

So I went with this solution:

- We remove a bucket from the map only after each update of the pipelines, using an event from the ConnectorsManager. I tried to have something resilient against race conditions.
- We improve the `get` and `has` functions to detect if the bucket is in use, as the Allocator only needs to detect buckets in use: either it's in `.buckets` (pending deployment or in use), or not (the bucket was removed from memory, pending removal from the connector).